File: src/javascript/algorithms/array/searchInsertPosition/searchInsertPosition.js
function searchInsert(nums, target) {
    let left = 0, right = nums.length - 1;
    while (left <= right) {
        const mid = Math.floor((left + right) / 2);
        if (nums[mid] === target) return mid;
        else if (nums[mid] < target) left = mid + 1;
        else right = mid - 1;
    }
    return left;
}

module.exports = searchInsert;

if (require.main === module) {
    const testCases = [
        { nums: [1, 3, 5, 6], target: 5, expected: 2 },
        { nums: [1, 3, 5, 6], target: 2, expected: 1 },
        { nums: [1, 3, 5, 6], target: 7, expected: 4 },
        { nums: [1, 3, 5, 6], target: 0, expected: 0 },
        { nums: [], target: 3, expected: 0 },
        { nums: [1], target: 0, expected: 0 },
        { nums: [1], target: 2, expected: 1 },
        { nums: [1], target: 1, expected: 0 },
    ];
    testCases.forEach((tc, i) => {
        const result = searchInsert(tc.nums, tc.target);
        console.log(`Test ${i+1}: searchInsert([${tc.nums}], ${tc.target}) = ${result} | Expected: ${tc.expected} | ${result === tc.expected ? 'PASS' : 'FAIL'}`);
    });
}

File: src/python/algorithms/array/searchInsertPosition/search_insert_position.py
def search_insert(nums, target):
    left, right = 0, len(nums) - 1
    while left <= right:
        mid = (left + right) // 2
        if nums[mid] == target:
            return mid
        elif nums[mid] < target:
            left = mid + 1
        else:
            right = mid - 1
    return left


if __name__ == "__main__":
    test_cases = [
        ([1, 3, 5, 6], 5, 2),
        ([1, 3, 5, 6], 2, 1),
        ([1, 3, 5, 6], 7, 4),
        ([1, 3, 5, 6], 0, 0),
        ([], 3, 0),
        ([1], 0, 0),
        ([1], 2, 1),
        ([1], 1, 0),
    ]
    for i, (nums, target, expected) in enumerate(test_cases, 1):
        result = search_insert(nums, target)
        print(f"Test {i}: search_insert({nums}, {target}) = {result} | Expected: {expected} | {'PASS' if result == expected else 'FAIL'}")
